### PR TITLE
feat(dashboard): add api for customizing the properties panel

### PR DESCRIPTION
**Pull request by jmbuss**
## Overview
Adds a configuration option in dashboard to customize the properties panel sections.

The property panel sections were moved from the sidePanel folder in components to the customization folder. In general, the only difference in those section components is that the interface has changed. Previously the interface was a widget type, and it has been changed to getters and setters fro the relevant properties on a widget. Some of the components have more extensive refactors like the thresholds section. This is to account for a new UX when displaying the section for multiple widgets.

## Legal
This project is available under the [Apache 2.0 License](http://www.apache.org/licenses/LICENSE-2.0.html).

### CODE REVIEW

File: packages/dashboard/src/components/internalDashboard/index.tsx [hunk 1]
```diff
@@ -253,7 +254,8 @@ const InternalDashboard: React.FC<InternalDashboardProperties> = ({ onSave, edit
             <WebglContext viewFrame={viewFrame} />
           </div>
         }
-        rightPane={<SidePanel messageOverrides={DefaultDashboardMessages} />}
+        rightPane={<SidePanel sections={propertiesSections ?? []} />}
+        // rightPane={<SidePanelOld messageOverrides={DefaultDashboardMessages} />}
```
Comment (mnischay) [inline packages/dashboard/src/components/internalDashboard/index.tsx:258]: you can probably remove this ?

File: packages/dashboard/src/customization/propertiesSection.tsx [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+type Setter<T, V> = (target: T, value: V) => T;
+
+type Lens<W> = <T>(selector: Getter<W, T>, updater: Setter<W, T>) => [Maybe<T>, (newValue: T) => void];
+
+type PropertyLens<W extends DashboardWidget> = Lens<W['properties']>;
+
+const NO_SIZE = { height: 0, width: 0 };
+const NO_POSITION = { x: 0, y: 0 };
+
+function compositeValue<T>(values: T[]): Maybe<T> {
```
Comment (mnischay) [inline packages/dashboard/src/customization/propertiesSection.tsx:31]: super nit: better to use arrow function?

File: packages/dashboard/src/customization/propertiesSection.tsx [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+      dispatch(
+        onMoveWidgetsAction({
+          widgets: selection,
+          vector,
+          complete: true,
+        })
+      ),
+  ];
+
+  const type = compositeValue(selection.map((w) => w.type));
```
Comment (mnischay) [inline packages/dashboard/src/customization/propertiesSection.tsx:78]: super nit: typically we avoid alphabets as variable names ?

File: packages/dashboard/src/customization/propertiesSection.tsx [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+};
+
+export type PropertiesSectionProps<W extends DashboardWidget> = {
+  isVisible: FilterPredicate<W>;
+  render: RenderSection<W>;
+};
+
+export type DashboardSelection = NonNullable<ReturnType<typeof useSelection>>;
+
+export const isDashboardWidget = (w: DashboardWidget): w is DashboardWidget => !!w;
```
Comment (mnischay) [inline packages/dashboard/src/customization/propertiesSection.tsx:117]: curious : is it ok if we dont use alphabets ? may be its a personal pref..

File: packages/dashboard/src/customization/propertiesSections/shared/selectOneWidget.tsx [hunk 5]
```diff
@@ -0,0 +1,8 @@
+import React from 'react';
+import { Box } from '@cloudscape-design/components';
+
+export const SelectOneWidget: React.FC = () => (
```
Comment (mnischay) [inline packages/dashboard/src/customization/propertiesSections/shared/selectOneWidget.tsx:4]: nit: given no props, do we need ```React.FC``` ?

File: packages/dashboard/src/customization/propertiesSections/textSettings/text.tsx [hunk 6]
```diff
@@ -57,86 +49,61 @@ const defaultMessages = {
...
+type TextSettingsProps = FontSettings & {
+  updateFontColor: (newValue: FontSetting<'fontColor'>) => void;
+  toggleBold: (newValue: FontSetting<'isBold'>) => void;
+  toggleItalic: (newValue: FontSetting<'isItalic'>) => void;
+  toggleUnderlined: (newValue: FontSetting<'isUnderlined'>) => void;
+  updateFontSize: (newValue: FontSetting<'fontSize'>) => void;
+};
 
+const TextSettings: FC<TextSettingsProps> = ({
+  fontColor = '#000000',
```
Comment (mnischay) [inline packages/dashboard/src/customization/propertiesSections/textSettings/text.tsx:64]: may be should use cloudscape design token ?

File: packages/dashboard/src/customization/propertiesSection.tsx [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+     * the value of the selector is only defined
+     * if every widget in the selection has the same value
+     * otherwise we can not pick a single value to represent
+     * the selection
+     */
+    const propertyValues = selection.map((w) => selector(w.properties));
+    const propertyValue = compositeValue(propertyValues);
+
+    return [
+      propertyValue,
```
[the remark this answers is not in the document]
Comment (jmbuss) [reply packages/dashboard/src/customization/propertiesSection.tsx:91]: The sections do determine this. `useSelection` returns a hook called `useProperty` That hook is used in each section to get / set those properties

File: packages/dashboard/src/components/propertiesPanel/index.tsx [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+import React from 'react';
+import { useSelection } from '~/customization/propertiesSection';
+
+import './index.css';
+
+type SidePanelProps = {
+  sections: React.ReactElement[];
+};
+
+const SidePanel: React.FC<SidePanelProps> = ({ sections }) => {
```
Comment (tracy-french) [inline packages/dashboard/src/components/propertiesPanel/index.tsx:11]: Why type `React.FC<Props>` and not type the props?

```ts
const SidePanel = ({ sections }: SidePanelProps) => { [..] };
```

File: packages/dashboard/src/customization/propertiesSection.tsx [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+import { Just, Maybe, Nothing } from '~/util/maybe';
+
+export type FilterPredicate<W extends DashboardWidget> = (
+  w: DashboardWidget,
+  index: number,
+  selection: DashboardWidget[]
+) => w is W;
+export type RenderSection<W extends DashboardWidget> = (
+  selection: NonNullable<ReturnType<typeof useSelection<W>>>
+) => React.ReactElement;
```
Comment (tracy-french) [inline packages/dashboard/src/customization/propertiesSection.tsx:19]: These and other types in this file could benefit from doc comments.

File: packages/dashboard/src/customization/propertiesSection.tsx [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+  /**
+   * selection filter does not apply to entire selection
+   * this means we cannot correctly narrow the selection type
+   */
+  if (selectedWidgets.length === 0 || !isEqual(selectedWidgets, filteredSelection)) return undefined;
+
+  const selection = filteredSelection;
+
+  /**
+   * TECH DEBT: getSelectionBox should never be null given the above check
```
Comment (tracy-french) [inline packages/dashboard/src/customization/propertiesSection.tsx:51]: Great comment!

File: packages/dashboard/src/customization/propertiesSections/axisSettings/index.tsx [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+
+import { PropertiesSection, PropertiesSectionProps } from '~/customization/propertiesSection';
+import { AxisSettings } from '~/customization/settings';
+import { DashboardWidget } from '~/types';
+import { maybeWithDefault } from '~/util/maybe';
+import AxisSection from './section';
+
+const axisWidgetTypes: readonly string[] = ['line-chart', 'scatter-chart', 'bar-chart', 'status-timeline'];
+// The <StatusTimeline /> widget does not support the Y axis
+const axisWidgetTypesThatSupportYAxis: readonly string[] = ['line-chart', 'scatter-chart', 'bar-chart'];
```
Comment (tracy-french) [inline packages/dashboard/src/customization/propertiesSections/axisSettings/index.tsx:11]: Nice usage of readonly!

File: packages/dashboard/src/customization/propertiesSections/propertiesAndAlarmsSettings/section.tsx [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+        {getComponents()}
+      </SpaceBetween>
+    </ExpandableSection>
+  );
+};
+type TablePropertiesAlarmsSectionProps = PropertiesAlarmsSectionProps & {
+  items: TableItem[] | undefined;
+  updateItems: (newValue: TableItem[] | undefined) => void;
+};
+export const TablePropertiesAlarmsSection: FC<TablePropertiesAlarmsSectionProps> = ({
```
Comment (tracy-french) [inline packages/dashboard/src/customization/propertiesSections/propertiesAndAlarmsSettings/section.tsx:135]: Formatting! :p

File: packages/dashboard/src/customization/propertiesSections/propertiesAndAlarmsSettings/section.tsx [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+            };
+          }
+          return asset;
+        })
+        .filter((asset) => asset.properties.length > 0) ?? [];
+
+    updateSiteWiseAssetQuery({ assets });
+  };
+
+export const GeneralPropertiesAlarmsSection: FC<PropertiesAlarmsSectionProps> = ({
```
Comment (tracy-french) [inline packages/dashboard/src/customization/propertiesSections/propertiesAndAlarmsSettings/section.tsx:54]: This component and file would benefit from comments.

File: packages/dashboard/src/customization/propertiesSections/shared/useExpandable.ts [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+}) => {
+  const [expanded, setExpanded] = useState(defaultExpanded);
+
+  useEffect(() => {
+    if (syncCloseWhen) {
+      setExpanded(false);
+    }
+  }, [syncCloseWhen]);
+
+  return [expanded, setExpanded] as const;
```
Comment (tracy-french) [inline packages/dashboard/src/customization/propertiesSections/shared/useExpandable.ts:18]: Nice!

File: packages/dashboard/src/customization/propertiesSections/shared/useExpandable.ts [hunk 15]
```diff
@@ -0,0 +1,19 @@
...
+  syncCloseWhen: boolean;
+  defaultExpanded?: boolean;
+}) => {
+  const [expanded, setExpanded] = useState(defaultExpanded);
+
+  useEffect(() => {
+    if (syncCloseWhen) {
+      setExpanded(false);
+    }
+  }, [syncCloseWhen]);
```
Comment (tracy-french) [inline packages/dashboard/src/customization/propertiesSections/shared/useExpandable.ts:16]: Is the effect necessary?

File: packages/dashboard/src/customization/propertiesSections/sizeAndPositionSettings/section.spec.tsx [hunk 16]
```diff
@@ -3,11 +3,11 @@ import { Provider } from 'react-redux';
...
-import { MOCK_KPI_WIDGET } from '../../../../../testing/mocks';
-import { BaseSettings } from './index';
-import { configureDashboardStore } from '../../../../store';
-import type { KPIWidget } from '../../../../customization/widgets/types';
-import type { DashboardState } from '../../../../store/state';
+import { MOCK_KPI_WIDGET } from '../../../../testing/mocks';
+import { configureDashboardStore } from '../../../store';
+import type { KPIWidget } from '../../widgets/types';
+import type { DashboardState } from '../../../store/state';
+import { SizeAndPositionConfiguration } from './index';
```
Comment (tracy-french) [inline packages/dashboard/src/customization/propertiesSections/sizeAndPositionSettings/section.spec.tsx:10]: How do we order our imports?

File: packages/dashboard/src/customization/propertiesSections/thresholdSettings/index.tsx [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+  thresholdsWithAnnotations.some((t) => t === w.type);
+const supportsContainsOperator = (w: DashboardWidget): w is ThresholdsWidget =>
+  thresholdsWithContainsOperator.some((t) => t === w.type);
+
+// there is no widget type overlap between the 2 lists
+const isThresholdWidgetSupportsContainsOperator = (w: DashboardWidget): w is ThresholdsWidget =>
+  isThresholdsWidget(w) && supportsContainsOperator(w);
+const isThresholdWidgetSupportsAnnotations = (w: DashboardWidget): w is ThresholdsWithAnnotationsWidget =>
+  isThresholdsWidget(w) && supportsAnnotations(w);
+
```
Comment (tracy-french) [inline packages/dashboard/src/customization/propertiesSections/thresholdSettings/index.tsx:31]: Do we need all these utilities for this component?

File: packages/dashboard/src/customization/propertiesSections/thresholdSettings/thresholdsSection.tsx [hunk 18]
```diff
@@ -0,0 +1,230 @@
+import React, { FC, MouseEventHandler } from 'react';
```
Comment (tracy-french) [inline packages/dashboard/src/customization/propertiesSections/thresholdSettings/thresholdsSection.tsx:1]: This file would benefit from comments. There's a lot going on here.

File: packages/dashboard/src/util/maybe.ts [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+export const isJust = <T>(m: Maybe<T>): m is Just<T> => m.type === MaybeType.Just;
+
+export const maybeWithDefault = <T>(defaultVal: T, m: Maybe<T>): T => {
+  switch (m.type) {
+    case MaybeType.Nothing:
+      return defaultVal;
+    case MaybeType.Just:
+      return m.value;
+  }
+};
```
Comment (tracy-french) [inline packages/dashboard/src/util/maybe.ts:35]: If you make a public util, treat it like an API and document its intended usage.

File: packages/dashboard/src/customization/propertiesSections/textSettings/index.tsx [hunk 20]
```diff
@@ -0,0 +1,89 @@
...
+import { PropertiesSection } from '~/customization/propertiesSection';
+import { TextWidget } from '~/customization/widgets/types';
+import { DashboardWidget } from '~/types';
+import { maybeWithDefault } from '~/util/maybe';
+import TextSettings from './text';
+import LinkSettings from './link';
+
+const isTextWidget = (w: DashboardWidget): w is TextWidget => w.type === 'text';
+
+export const TextSettingsConfiguration: React.FC = () => (
```
Comment (tracy-french) [inline packages/dashboard/src/customization/propertiesSections/textSettings/index.tsx:12]: Looks clean!

**Review by tracy-french — CHANGES_REQUESTED** [review]
Great work, @jmbuss! This is a foundational change for the dashboard and I think you nailed it. Before we merge this in, let's add plenty of in-source documentation. It's complex, yet fundamental. Let's make it super easy to build with. :)

**Review by corteggiano — APPROVED** [review]

**Review by tracy-french — APPROVED** [review]
Fantastic work @jmbuss! Thank you for increasing the extensibility of the dashboard with your contribution!

**Pull request merged**